pkg/pillar/cmd/zedagent/zedagent.go

```diff
@@ -256,17 +269,150 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject) in
 
 	log.Functionf("Starting %s", agentName)
 
+	// Initialize zedagent context.
+	zedagentCtx := newContext()
+	zedagentCtx.ps = ps
+	zedagentCtx.hangFlag = *hangPtr
+	zedagentCtx.fatalFlag = *fatalPtr
+
 	flowlogQueue := make(chan *flowlog.FlowMessage, flowlogQueueCap)
 	triggerDeviceInfo := make(chan struct{}, 1)
 	triggerHwInfo := make(chan struct{}, 1)
 	triggerObjectInfo := make(chan infoForObjectKey, 1)
-	zedagentCtx := zedagentContext{
-		ps:                ps,
-		FlowlogQueue:      flowlogQueue,
-		TriggerDeviceInfo: triggerDeviceInfo,
-		TriggerHwInfo:     triggerHwInfo,
-		TriggerObjectInfo: triggerObjectInfo,
-		zedcloudMetrics:   zedcloud.NewAgentMetrics(),
+	zedagentCtx.FlowlogQueue = flowlogQueue
+	zedagentCtx.TriggerDeviceInfo = triggerDeviceInfo
+	zedagentCtx.TriggerHwInfo = triggerHwInfo
+	zedagentCtx.TriggerObjectInfo = triggerObjectInfo
+
+	// Initialize pubsub channels.
+	initializePublications(zedagentCtx)
+	initializeSubscriptions(zedagentCtx)
+
+	// upgradeconverter ensures we have a ConfigItemValueMap so we
+	// read it to get the initial values
+	item, err := zedagentCtx.pubGlobalConfig.Get("global")
+	if err != nil {
+		log.Fatalf("ConfigItemValueMap missing: %s", err)
+	}
+	zedagentCtx.globalConfig = item.(types.ConfigItemValueMap)
+	log.Functionf("initialized GlobalConfig")
+
+	// Run a periodic timer so we always update StillRunning
+	stillRunning := time.NewTicker(25 * time.Second)
+	ps.StillRunning(agentName, warningTime, errorTime)
+
+	initializeDirs()
+
+	// Apply saved radio config ASAP.
```

Comment:
This seems like a reasonable place to activate all the subscriptions; we have the Debug setting from GlobalConfig and AFAICT we have the contexts set up.

Reply:
Yeah it seems that subscriptions can be all created and at the same time activated after we have GlobalConfig (so no separate Activate call needed; GlobalConfig comes from publication).
That's the main issue with the current Run() function - it goes through 3 event loops and publications and subscriptions are created and activated in different places without commenting if there are any dependencies, intentions to create/activate them later, etc.

Reply:
The old approach was to 1) setup publications before subscriptions and 2) always activate after each NewSubscription.
That way less likely to forget any activate. (With this PR one has to remember to add that separately located Activate call in addition to adding the subscription and using it in the event loop(s))

But perhaps the concern which lead to the separate Activate call was based on a misunderstanding on my part.
We need to make sure that the state (in the contexts and elsewhere) is set up before the handle functions are called.
But the handle functions are not called until an event loop reads from the channel; Activate just makes pubsub post an event on the channel.
So it would be worth while to try getting rid of Activate i.e., threat it as if Activate is set in the SubscriptionOptions and see if we have any issues.

Reply:
As I remember persist subscriptions have different behaviour: https://github.com/lf-edge/eve/blob/dee0c391e23f40e08b7e7eacf2e532c03086c846/pkg/pillar/pubsub/subscribe.go#L73-L103

Reply:
I need to have a chat with the person who wrote that code ;-)

Seriously, I wonder if we should change that to also rely on the MsgChan to make it be uniform and then get rid of Activate.

Reply:
Upon closer inspection it turned out that all subscriptions (**incl. persisted ones**) can be activated once the zedagent context is initialized and device is onboarded, aka device UUID is known. This allows to simplify the code quite a bit and avoid having these Activate() calls all over the place.